diffmpm/particle.py

```diff
@@ -1,336 +1,568 @@
-
-
-@register_pytree_node_class
-class Particles:
-    """Container class for a set of particles."""
-
-    def __init__(
-        self,
-        loc: jnp.ndarray,
-        material: Material,
-        element_ids: jnp.ndarray,
-        initialized: bool = None,
-        data: Tuple[jnp.ndarray, ...] = None,
-    ):
-        """
-        Initialize a container of particles.
-
-        Arguments
-        ---------
-        loc: jax.numpy.ndarray
-            Location of the particles. Expected shape (nparticles, 1, ndim)
-        material: diffmpm.material.Material
-            Type of material for the set of particles.
-        element_ids: jax.numpy.ndarray
-            The element ids that the particles belong to. This contains
+from jax import jit, lax, vmap
+from jax.tree_util import register_pytree_node_class, Partial
+from jax.typing import ArrayLike
+
+from diffmpm.element import _ElementsState
+from diffmpm.materials import _Material
+
+import chex
+
+
+@chex.dataclass(frozen=True)
+class _ParticlesState:
+    nparticles: int
+    loc: chex.ArrayDevice
+    material: _Material
+    element_ids: chex.ArrayDevice
+    mass: chex.ArrayDevice
+    density: chex.ArrayDevice
+    volume: chex.ArrayDevice
+    size: chex.ArrayDevice
+    velocity: chex.ArrayDevice
+    acceleration: chex.ArrayDevice
+    momentum: chex.ArrayDevice
+    strain: chex.ArrayDevice
+    stress: chex.ArrayDevice
+    strain_rate: chex.ArrayDevice
+    dstrain: chex.ArrayDevice
+    f_ext: chex.ArrayDevice
+    traction: chex.ArrayDevice
+    reference_loc: chex.ArrayDevice
+    dvolumetric_strain: chex.ArrayDevice
+    volumetric_strain_centroid: chex.ArrayDevice
+    state_vars: dict
+
+
+def init_particle_state(
+    loc: chex.ArrayDevice,
+    material: _Material,
+    element_ids: chex.ArrayDevice,
+    init_vel: chex.ArrayDevice = 0,
+):
+    """Initialize a container of particles.
+
+    Parameters
+    ----------
+    loc: ArrayLike
+        Location of the particles. Expected shape (nparticles, 1, ndim)
+    material: diffmpm.materials._Material
+        Type of material for the set of particles.
+    element_ids: ArrayLike
+        The element ids that the particles belong to. This contains
         information that will make sense only with the information of
         the mesh that is being considered.
-        initialized: bool
-            False if particle property arrays like mass need to be initialized.
-        If True, they are set to values from `data`.
-        data: tuple
-            Tuple of length 13 that sets arrays for mass, density, volume,
+    initialized: bool
+        `False` if particle property arrays like mass need to be initialized.
+        If `True`, they are set to values from `data`.
+    data: tuple
+        Tuple of length 13 that sets arrays for mass, density, volume,
         velocity, acceleration, momentum, strain, stress, strain_rate,
         dstrain, f_ext, reference_loc and volumetric_strain_centroid.
```

Comment:
These parameters are not present in this method also docstring for init_vel can be added.